### DISCUSSION TITLE
how to toggle an output gpis?

### DISCUSSION THREAD

**Discussion by billy27607** [RavenSystem/esp-homekit-devices#1759] [MEPLHAA Scripts]
I have been working on these Treatlife 3-way switches. They are cool because they have power sensor that allows them to be used with dumb SPDT switch on the other side. The relay on the smart switch is also SPDT.

I have a config (see below) that works, sort of. Note that the actual state of the load is reported by the power sensor on pin 14. 

The problem is that this config is dependent on which way the travelers on the switch were wired in the wall. You have a 50/50 chance of getting it right. If you not get it right, invert the output values of "0" and "1" and it all works.

But the fact that I have to change the code dependent on how the switch is installed is seriously tweaking my techie OCD. What I need to be able to do is config so that the "0" and "1" actions toggle the state of the relay rather than setting it to a specific value. Since the state is set by the power sensor this works,

The question is: How do I create actions "0" and "1" that both do the same thing, toggle the state of the load. I am not terribly familiar with this firmware and I have not been able to see how to do it.

Does anyone have any suggestions or do I have to config for each install?

{
    "c": {
        "l": 5,
        "I": 1,
        "b": [{
            "g": 13,
            "t": 5
        }]
    },
    "a": [{
        "t": 1,
        "s": 5,
        "0": {
            "r": [{
                "g": 12,
                "v": 0
            }]
        },
        "1": {
            "r": [{
                "g": 12,
                "v": 1
            }]
        },
        "b": [{
            "g": 13
        }],
        "g0": [{
            "g": 14,
            "t": 1
        }],
        "g1": [{
            "g": 14,
            "t": 0
        }]
    }]
}

**Comment by andyallcorn** [comment]
I’m not entirely clear on which part of this config needs to be reversed sometimes?

1. The status inputs on gpio 14
2. The binary input on gpio 13
3. The outputs on gpio 12

It _sounds_ a bit like you’re swapping the outputs (3), but possibly only because the switch (2) is the other way around?

Would setting the switch to trigger a state change whenever you use it in either direction achieve what you want here? (I _think_ changing the buttons to ```"b": [{"g": 13}, {"g": 13, "t": 0}]``` does that.) Apologies if I’ve misunderstood.

**Reply by billy27607** [reply]
Andy, thanks for your help and certainly any confusion is on me. 

I am inverting the output on pin 12 to make it work. Both the physical button and the button in the app behave the same, so I do not think that I can fix it by changing the button. In fact, I have tried exactly what you suggested.

I think that the secret is in the "0" and "1" actions. As written each of these statements write a specific value to the output gpio. To make it work without tweaking the code based on how the travelers are wired, I need each of those statements to toggle the output rather than set it to a specific state. Both actions would do the same thing and toggle.

I know that will work, because I coded the whole thing up in C before. But this firmware is really nice and much easier to manage than my custom code. 

I just do not know how to configure that to make it happen. Is there a way to set the value of an output based on the current value of the output? Something like:

"0":{"r":[{"g":12,"v":!v}]}
"1":{"r":[{"g":12,"v":!v}]}

Again, thanks for helping a neophyte figure this out.

**Reply by andyallcorn** [reply]
Ah, okay. I don't _think_ you can do exactly that directly, but you could certainly use another switch (hidden from HomeKit, but exposed to the HAA Manager app, with "h":2) that does the actual gpio actions, and then send a "toggle state" message _to that switch_ as both of the actions in your primary switch. Make sense? Something like the below JSON (which I haven't tested...).

Though, I would think you still need to account for initial state somewhere, so won't you need at least one hard-coded value...?
```{
	"c": {
		"l": 5,
		"I": 1,
		"b": [{
			"g": 13,
			"t": 5
		}]
	},
	"a": [{
		"t": 1,
		"s": 5,
		"0": {
			"m": [
				[2, 4]
			]
		},
		"1": {
			"m": [
				[2, 4]
			]
		},
		"b": [{
			"g": 13
		}],
		"g0": [{
			"g": 14,
			"t": 1
		}],
		"g1": [{
			"g": 14,
			"t": 0
		}],
		"es": [{
			"t": 1,
			"s": 5,
			"h": 2,
			"0": {
				"r": [{
					"g": 12,
					"v": 0
				}]
			},
			"1": {
				"r": [{
					"g": 12,
					"v": 1
				}]
			}
		}]
	}]
}

**Reply by billy27607** [reply]
That is a really good idea. I am away from my shop for a couple of days and will not be able to test it, but will put up the result as soon as I can.

As for the initial state, if it restores the last previous state it will probably be good enough. 

As before thanks for your time...

**Reply by billy27607** [reply]
OK.. I tried it and it fixed the problem with the buttons, Both the physical one on the switch and in the home app. BUT, somehow it broke the two state functions reading the power sensor and setting the state back to home kit. That means that if you use the smart switch button or the home app it works, but if you change the dumb switch on the other end of the traveler the state does not get updated in HomeKit and things get backwards. 

II cannot figure why this approach broke the power sensor reading... I guess that I will just keep setting the "0" and "1" actions after the switch is installed.

**Reply by billy27607** [reply]
I have refined my testing approach and the code that @andyallcorn submitted worked well. The working code is below for anyone who would like to use these switches. They work very well..

`{
    "c":{
        "l":5,
        "I":1,
        "b":[{"g":13,"t":5}]
        },
    "a":[{
        "t":1,
        "s":5,
        "0":{"m":[[2,4]]},
        "1":{"m":[[2,4]]},
        "b":[{"g":13}],
        "g0":[{"g":14,"t":1}],
        "g1":[{"g":14,"t":0}],
        "es":[{
            "t": 1,
            "s": 5,
            "h": 2,
            "0":{"r":[{"g":12,"v":0}]},
            "1":{"r":[{"g":12,"v":1}]}
        }]
    }]
}`

**Reply by andyallcorn** [reply]
Ah, amazing! Pleased you got it working. (I was rather out of ideas!)

**Comment by billy27607** [comment]
Thanks for all of your help. I am not sure why the first load of your code did not work, but I went back to it and got things working. It makes for a really nice and low cost device.

Thanks again.


Billy
